### PR TITLE
[Draft/RFC] Envoy + ext_proc alternative for KEP-NNNN

### PR DESCRIPTION
**Pull request by mastersingh24**
**Status:** draft / RFC. Posted for discussion alongside #838 (from-scratch Go router). Not asking for review-to-merge — looking for a steer on whether the Envoy + ext_proc shape is interesting enough to pursue as v2 or post-v1 evolution.

## What this is

An alternative implementation of [KEP-NNNN](https://github.com/kubernetes-sigs/agent-sandbox/pull/758) that keeps the KEP's three core mechanisms (informer-backed UID→PodIP map, direct Pod IP dispatch, drift handling) in Go but delegates **everything else** (TLS, mTLS, rate limit, circuit breaker, retries, access logs, tracing, WAF, JWT) to Envoy.

The Go service is a single ext_proc gRPC server (~300 lines + cache) that owns the K8s-aware state. Envoy handles the data plane.

## Architecture

```
client ──► Envoy listener (:8080)
            │
            ├── HTTP filter: ext_proc
            │     │
            │     │  gRPC ProcessingRequest (RequestHeaders)
            │     ▼
            │   ext-proc-sandbox-router service
            │     │  - K8s Pod informer (labeled sandbox-name-hash)
            │     │  - In-memory UID → PodIP map
            │     │  - Reads X-Sandbox-UID + X-Sandbox-ID etc.
            │     │  - Returns header mutation:
            │     │    x-envoy-original-dst-host = <PodIP>:<port>
            │     │    (or DNS form on cache miss)
            │     ▼
            │   gRPC ProcessingResponse (HeaderMutation + ClearRouteCache)
            │
            └── HTTP filter: router → ORIGINAL_DST cluster
                                       (reads x-envoy-original-dst-host)
                ─────► sandbox Pod
```

Per-request cost: one in-cluster gRPC round-trip (~sub-ms) plus the Pod-IP dispatch.

## KEP requirements → status

| KEP item | Status | Notes |
|---|---|---|
| Informer watching sandbox Pods | ✅ | Same `cache` code as the from-scratch router |
| UID → PodIP in-memory map | ✅ | Shared between both prototypes |
| Direct dispatch, bypass Service | ✅ | Envoy ORIGINAL_DST cluster reads `x-envoy-original-dst-host` |
| `X-Sandbox-UID` request shape | ✅ | |
| Active cache invalidation on connection error | ✅ | Envoy outlier detection ejects 5xx-returning Pod IPs |
| Hybrid DNS fallback (cache miss) | ✅ | ext_proc returns DNS form when UID isn't cached |
| Sync period | ✅ | Standard 10m informer resync |
| Strict input validation | ✅ | Performed in ext_proc handler |
| TokenReview integration | ❌ | Deferred — drops in as an `ext_authz` filter or sibling ext_proc, not yet wired |
| TLS termination / mTLS | ✅✅ | Envoy native (the from-scratch router has to implement this) |

## Comparison with the from-scratch Go router (#838)

| Concern | From-scratch Go (#838) | Envoy + ext_proc (this) |
|---|---|---|
| TLS / mTLS | Custom (~400 LOC + fsnotify hot-reload) | Envoy native |
| Rate limit | Not implemented | `local_ratelimit` filter (config-only) |
| Circuit breaker | Not implemented | Envoy outlier detection (built in) |
| Retries | Custom retry transport (~150 LOC) | Envoy retry policy + hedged retries |
| Access logs | Custom middleware (~150 LOC) | Envoy access log (config-only) |
| Tracing | Custom OTel middleware (~100 LOC) | Envoy OTel native |
| Custom Go LOC | ~5,000 | ~300 |
| K8s API surface | Same informer | Same informer |
| Operational footprint | One Go binary | Envoy + Go service (two Deployments) |
| TokenReview | Implemented (#838) | Deferred |

## Validated

- Unit tests for cache + ext_proc handler.
- Standalone load harness at `cmd/load-test/` exercising real `Server` + real `Cache` (fake K8s client) — single replica sustains **~38k req/s** at 5k sandboxes with 80% cache hit, no errors, p50 1.4ms / p99 6.0ms. Bottleneck is gRPC machinery, not the cache.
- End-to-end run on kind: Envoy + ext_proc Deployments come up, `/healthz` returns 200, real sandbox Pods get routed via the cache, deleting a sandbox triggers outlier ejection within a few requests.

## Open questions

1. **Is the two-process operational model acceptable?** Trades ~5k LOC of Go for ~200 lines of Envoy YAML, but adds Envoy as a hard runtime dep. For OSS Kubernetes users without an existing Envoy footprint, this may be the wrong default. For users running a service mesh, it's nearly free.
2. **Should this replace #838 or coexist?** Two natural positions: (a) #838 ships as v1 (simpler operational story, no Envoy dep), this ships as v2 / opt-in for users who want the enterprise feature set; (b) this is the long-term answer and #838 is the bridge while Envoy adoption catches up.
3. **TokenReview path.** Two options here: an `ext_authz` filter pointing at a small authn extension of this same service (clean, separates concerns), or a second ext_proc filter sharing the same Go process (one fewer Deployment). Either works; preferences welcome.
4. **UID source.** Same question as #838 — the KEP says "Sandbox CRD UID (read from Pod labels)" but the controller doesn't stamp UID as a label. Both prototypes read it from `Pod.metadata.ownerReferences`. Want me to file a small controller change to also stamp `agents.x-k8s.io/sandbox-uid`?

Happy to close this out if the direction is clearly "stay all-Go" — the work was worth it as a comparison data point either way.

### CODE REVIEW

File: sandbox-router/internal/extproc/server.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+	}
+	return hasAlphanum
+}
+
+// joinHostPort formats "host:port" without the IPv6 bracket dance —
+// our cache stores IPv4 addresses or DNS names, never bare IPv6 strings.
+// (Pod IPv6 support would need updating here.)
+func joinHostPort(host string, port int) string {
+	return host + ":" + strconv.Itoa(port)
+}
```
[the remark this answers is not in the document]
Comment (mastersingh24) [reply sandbox-router/internal/extproc/server.go]: Fixed in b841c55. Swapped `joinHostPort` to use `net.JoinHostPort` (which brackets IPv6 literals per RFC 3986) and updated the comment to call out *why* — sandbox Pods on dual-stack or IPv6-only clusters have a bare IPv6 string in `Pod.Status.PodIP`, and an unbracketed value would make Envoy reject the `x-envoy-original-dst-host` header.

Added table tests for v4, DNS, loopback v6, link-local v6, and a full v6 address, plus a `TestResolve_IPv6PodIPGetsBracketed` end-to-end through `resolve()` so we lock in the cache → bracketed-target path. Applied locally (no Copilot co-author / CLA wrinkle).

Heads-up: the from-scratch Go router on #838 has the same bug in `sandbox-router/proxy/resolve.go` and `target.go` (`host + ":" + strconv.Itoa(t.Port)`). I'll fix that there in a follow-up commit on that branch.

File: sandbox-router/internal/extproc/server.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+		return &extprocv3.ProcessingResponse{
+			Response: &extprocv3.ProcessingResponse_RequestHeaders{
+				RequestHeaders: &extprocv3.HeadersResponse{
+					Response: &extprocv3.CommonResponse{Status: extprocv3.CommonResponse_CONTINUE},
+				},
+			},
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (mastersingh24) [reply sandbox-router/internal/extproc/server.go:201]: You're right — fixed in c4d8aa0. After pulling the ext_proc spec to be sure: the oneof variant is the phase tag, not just an envelope, and replying with the wrong variant aborts the stream. Our previous default branch always wrapped in `ProcessingResponse_RequestHeaders`, so a misconfigured `processing_mode` that sent us, say, `RequestBody` would have produced a failed user request rather than the benign no-op the comment claimed.

Replaced with a `continueFor(req)` helper that switches on the request's oneof and returns a phase-matched CONTINUE: `HeadersResponse` / `BodyResponse` for the four header/body phases (both carry a `CommonResponse`), empty `TrailersResponse` for the two trailer phases (no `CommonResponse` field — absent `HeaderMutation` is the equivalent of CONTINUE). Unknown future oneof variants return nil and `Process` drops them rather than sending a wrong-phase reply.

`TestHandle_PhaseMatchedContinue` exercises all five non-RequestHeaders phases; `TestHandle_UnknownPhaseReturnsNil` locks in the nil contract. Applied locally, no Copilot co-author.

File: sandbox-router/cmd/ext-proc/main.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+	grpcSrv := grpc.NewServer()
+	extprocv3.RegisterExternalProcessorServer(grpcSrv, srv)
+
+	// Health server — NOT_SERVING until informer.HasSynced(); Envoy's
+	// gRPC health check sees this and routes around us until ready.
+	healthSrv := health.NewServer()
+	healthSrv.SetServingStatus(healthServiceName, healthpb.HealthCheckResponse_NOT_SERVING)
+	healthSrv.SetServingStatus("", healthpb.HealthCheckResponse_NOT_SERVING)
+	healthpb.RegisterHealthServer(grpcSrv, healthSrv)
```
[the remark this answers is not in the document]
Comment (mastersingh24) [reply sandbox-router/cmd/ext-proc/main.go]: Right again — fixed in f51b180. The Deployment's livenessProbe comment said "tracks process-aliveness rather than informer-readiness" but the code defeated that intent by gating BOTH the named and the default services on informer sync. On a large cluster where initial LIST exceeds the liveness window, the kubelet would restart-loop the pod before sync ever completed.

Split the two services by design:

- `""` (default, kubelet livenessProbe target): SERVING from the moment the gRPC server is registered, period. Only flipped on process exit, never on informer sync.
- `envoy.service.ext_proc.v3.ExternalProcessor` (readiness-gating, kubelet readinessProbe + Envoy gRPC health target): NOT_SERVING until `WaitForSync` returns, SERVING after, NOT_SERVING again on shutdown so Envoy + the kubelet stop routing while in-flight streams drain.

Also reviewed the shutdown path: on SIGTERM we now flip only the named service to NOT_SERVING — keeping the default SERVING so the kubelet livenessProbe doesn't trip mid-drain and SIGKILL us before GracefulStop finishes.

Extracted into a `newHealthServer()` helper with a test (`TestHealthServerSplitsLivenessAndReadiness`) pinning `default == SERVING, named == NOT_SERVING` at boot — so a future refactor can't silently re-couple them. Applied locally.

File: sandbox-router/internal/extproc/server.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+func (s *Server) onRequestHeaders(_ context.Context, hdrs *extprocv3.HttpHeaders) *extprocv3.ProcessingResponse {
+	r := readHeaders(hdrs.Headers)
+
+	if r.id == "" {
+		return immediate(400, `{"detail":"X-Sandbox-ID header is required."}`)
+	}
+	if !validNamespace(r.namespace) {
+		return immediate(400, `{"detail":"Invalid namespace format."}`)
+	}
+	if r.port == 0 {
```
[the remark this answers is not in the document]
Comment (mastersingh24) [reply sandbox-router/internal/extproc/server.go]: Fixed in 6437679. Half-right: `r.port < 1` was already catching negatives (the `|| r.port == 0` was redundant with that), but you're right that >65535 was sailing through and getting forwarded into `x-envoy-original-dst-host` as e.g. `10.0.0.1:99999`, which Envoy rejects with a less actionable error than our 400.

Tightened to `r.port < 1 || r.port > 65535`. Converted the existing single-case test into a table covering non-numeric, empty, zero, negative, way-negative, just-over-max, big number, and max-int32 — plus a `TestHandle_PortBoundariesAccepted` so the legal boundaries (1 and 65535) don't accidentally get caught later.

Same gap exists on the from-scratch Go router (#838, `proxy/headers.go`); I'll fix it there in a follow-up.

**Comment by mastersingh24** [comment]
Parity with the WebSocket / browser-backend fix on #838 (d04bfb5) landed here in b46cb13, in the shapes that fit the ext_proc design:

1. **`Origin` stripping on upgrade — in the Go handler.** When `readHeaders` sees both `Connection: Upgrade` and a non-empty `Upgrade` header (same predicate `httputil.ReverseProxy` uses internally), the `HeaderMutation` returned to Envoy gains `RemoveHeaders: ["origin"]` alongside the existing dst-host `SetHeaders`. Backends that validate `Origin == Host` for CSRF (vscode-server, Jupyter) no longer reject the upgrade with a `1006` close. Non-upgrade traffic preserves `Origin` so CORS preflights still work.

2. **`X-Forwarded-Host` — in the Envoy config.** One-liner `request_headers_to_add` at the virtual_host level with value `%REQ(:AUTHORITY)%`. `X-Forwarded-For` / `-Proto` were already free via `use_remote_address: true`; `-Host` doesn't come with that setting and needs to be wired separately.

Tests: `TestHandle_StripsOriginOnUpgrade` and `TestHandle_NonUpgradePreservesOrigin` cover the mutation side; `TestReadHeaders_UpgradeDetection` is a table covering the 8 corner cases of the upgrade predicate (both headers required, Connection token-list parsing, case-insensitivity, etc.).

**Review by aditya-shantanu — COMMENTED** [review]
On the RFC question of whether the Envoy + ext_proc shape is sound — the architecture reads well, and not exposing an `X-Sandbox-Pod-IP` override (cache/DNS only) avoids a class of SSRF the from-scratch router in #838 has to defend against. Two security-relevant points for the design discussion:

**1. `x-envoy-original-dst-host` trust boundary (deploy/envoy/configmap.yaml).**
The `sandbox_upstream` ORIGINAL_DST cluster uses `use_http_header: true`, so the entire routing decision is "whatever this header says". For normal traffic the ext_proc filter overwrites it (`OVERWRITE_IF_EXISTS_OR_ADD`) and `failure_mode_allow: false` keeps it fail-closed, which is the right call. But there is no `request_headers_to_remove: [x-envoy-original-dst-host]` on the listener/route to strip a *client-supplied* value on ingress before ext_proc runs. As defense-in-depth for `use_http_header`, recommend stripping any inbound `x-envoy-original-dst-host` at the edge so the ext_proc handler is provably the only writer of that header — otherwise the security of the whole data path rests on ext_proc never being accidentally disabled on a route (e.g. the way `/healthz` already disables it). This is worth nailing down in the RFC since it is the central security property of the Envoy variant.

**2. `X-Sandbox-ID` is not validated before DNS interpolation (internal/extproc/server.go).**
`onRequestHeaders` validates namespace (`validNamespace`) and port, but `r.id` flows straight into the DNS form `r.id + "." + r.namespace + ".svc." + clusterDomain` in `resolve()` with no checks. The Python router this is meant to be compatible with runs `sandbox_id` through a DNS-label regex specifically to prevent DNS-injection / traversal-style inputs. Suggest applying the same DNS-label check to `id`. (Same gap exists in #838 — flagged there too.)

Neither blocks the RFC direction; both are about making the security story airtight if this path is pursued.

**Comment by mastersingh24** [comment]
@aditya-shantanu fixed in b23a65b:

- **`x-envoy-original-dst-host` ingress strip** → new `envoy.filters.http.header_mutation` filter runs *before* ext_proc in the HCM chain and removes any client-supplied value. ext_proc still always sets the header, so it's provably the only writer. A future route that disables ext_proc gets 503 (fail-closed) rather than dispatching to whatever the client asked for.
- **`X-Sandbox-ID` DNS-label** → same `validDNSLabel` as #838, applied to ID and namespace.
- **`Authorization` strip** (audit found this) → added `"authorization"` to the always-present `HeaderMutation.RemoveHeaders`. Test covers both upgrade and non-upgrade paths.

**Review by aditya-shantanu — COMMENTED** [review]
Re-reviewed at head 6a82e06. The only change since my previous review (b46cb13) is a merge of main — none of the PR's own sandbox-router files changed, so earlier feedback stands as-is. The branch is currently conflicting with main and CI is red; this needs a rebase before it can progress.
